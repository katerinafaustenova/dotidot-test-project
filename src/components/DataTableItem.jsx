import { formatDistance } from "date-fns";

const DataTableItem = ({
  dataSource,
  column,
  editMode,
  archived,
  setArchived,
}) => {
  if (!dataSource || !column) return;
  const getItemData = () => {
    switch (column) {
      case "archived":
        return (
          <select
            id={`${dataSource.name}-${column}`}
            name="archived"
            value={archived}
            onChange={(e) => setArchived(e.target.value === "true")}
            disabled={!editMode}
          >
            <option value="true">True</option>
            <option value="false">False</option>
          </select>
        );
      case "createdAt":
        const createdAtDate = new Date(dataSource[column]);
        const localeDate = createdAtDate.toLocaleDateString();
        return localeDate;
      case "updatedAt":
      case "lastImport":
        const now = new Date();
        const updatedImportedDate = new Date(dataSource[column]);
        const distance = formatDistance(updatedImportedDate, now);
        return `${distance} ago`;
      default:
        return dataSource[column];
    }
  };
  return <td key={`${dataSource.id}-${column}`}>{getItemData()}</td>;
};

export default DataTableItem;
