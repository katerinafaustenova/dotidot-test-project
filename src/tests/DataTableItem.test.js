import { render } from "@testing-library/react";
import { formatDistance } from "date-fns";
import renderer from "react-test-renderer";
import DataTableItem from "../components/DataTableItem";

describe("DataTableItem", () => {
  const dataProvider = [
    { description: "no props", data: {} },
    {
      description: "all props",
      data: {
        dataSource: {
          id: "1",
          name: "Test Name",
          archived: false,
          createdAt: "2023-11-28",
          updatedAt: "2023-12-01",
          icon: "icon",
          itemsCount: 5,
          lastImport: "2023-12-01",
        },
        column: "archived",
        editMode: false,
        archived: "false",
        setArchived: jest.fn(),
      },
    },
    {
      description: "updatedAt prop empty",
      data: {
        dataSource: {
          id: "1",
          name: "",
          archived: false,
          createdAt: "2023-11-28",
          updatedAt: "",
          icon: "icon",
          itemsCount: 5,
          lastImport: "2023-12-01",
        },
        column: "updatedAt",
        editMode: false,
        archived: "false",
        setArchived: jest.fn(),
      },
    },
    {
      description: "createdAt prop missing",
      data: {
        dataSource: {
          id: "1",
          name: "",
          archived: false,
          updatedAt: "",
          icon: "icon",
          itemsCount: 5,
          lastImport: "2023-12-01",
        },
        column: "createdAt",
        editMode: false,
        archived: "false",
        setArchived: jest.fn(),
      },
    },
    {
      description: "column prop missing",
      data: {
        dataSource: {
          id: "1",
          name: "Test Name",
          archived: false,
          updatedAt: "",
          icon: "icon",
          itemsCount: 5,
          lastImport: "2023-12-01",
        },
        editMode: false,
        archived: "false",
        setArchived: jest.fn(),
      },
    },
    {
      description: "whole dataSource prop missing",
      data: {
        column: "lastImport",
        editMode: false,
        archived: "false",
        setArchived: jest.fn(),
      },
    },
    {
      description: "different itemsCount prop type",
      data: {
        dataSource: {
          id: "1",
          name: 12345,
          archived: false,
          updatedAt: "",
          icon: "icon",
          itemsCount: "10",
          lastImport: "2023-12-01",
        },
        column: "itemsCount",
        editMode: false,
        archived: "false",
        setArchived: jest.fn(),
      },
    },
  ];

  dataProvider.forEach((props) => {
    it(`renders correctly with ${props.description}`, () => {
      const tree = renderer.create(<DataTableItem {...props.data} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});

describe("getItemData fn", () => {
  const dataProvider = [
    {
      description: "full text string format",
      data: {
        dataSource: {
          updatedAt: "2023-12-01T17:41:18+01:00",
        },
      },
    },
    {
      description: "ISO format",
      data: {
        dataSource: {
          updatedAt: "2023-12-01",
        },
      },
    },
    {
      description: "short date format",
      data: {
        dataSource: {
          updatedAt: "12/01/2023",
        },
      },
    },
    {
      description: "long date format",
      data: {
        dataSource: {
          updatedAt: "01 Dec 2023",
        },
      },
    },
  ];

  dataProvider.forEach((props) => {
    it(`renders formatted updatedAt date correctly with ${props.description}`, () => {
      const { getByText } = render(
        <table>
          <tbody>
            <tr>
              <DataTableItem
                dataSource={props.data.dataSource}
                column="updatedAt"
              />
            </tr>
          </tbody>
        </table>
      );

      const updatedAtDate = new Date(props.data.dataSource.updatedAt);
      const updatedAtDistance =
        formatDistance(updatedAtDate, new Date()) + " ago";

      expect(getByText(updatedAtDistance)).toBeInTheDocument();
    });
  });
});
