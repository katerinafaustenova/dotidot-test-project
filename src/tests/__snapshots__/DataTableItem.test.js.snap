// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTableItem renders correctly with all props 1`] = `
<td>
  <select
    disabled={true}
    id="Test Name-archived"
    name="archived"
    onChange={[Function]}
    value="false"
  >
    <option
      value="true"
    >
      True
    </option>
    <option
      value="false"
    >
      False
    </option>
  </select>
</td>
`;

exports[`DataTableItem renders correctly with column prop missing 1`] = `null`;

exports[`DataTableItem renders correctly with createdAt prop missing 1`] = `
<td>
  --
</td>
`;

exports[`DataTableItem renders correctly with different itemsCount prop type 1`] = `
<td>
  10
</td>
`;

exports[`DataTableItem renders correctly with no props 1`] = `null`;

exports[`DataTableItem renders correctly with updatedAt prop empty 1`] = `
<td>
  --
</td>
`;

exports[`DataTableItem renders correctly with whole dataSource prop missing 1`] = `null`;
